### ISSUE TITLE
nsqd: is a bug of "func (n *NSQD) Main() error" ?

### ISSUE THREAD

**Issue by hoojos** [nsqio/nsq#1229]
Hello:
    I'm not a native English speaker, please forgive my grammar error.  

  I'm confused about the code about `nsq/nsqd.go: line 246~285`. 
this follow code will lead to only the error of first call `exitFunc` be pushed to `exitCh`, if first time is ok, but next time error is not nil, the next error will not be pushed to   `exitCh`,  why not only execute `once.Do` while `err != nil `
```

exitFunc := func(err error) {
		once.Do(func() {
			if err != nil {
				n.logf(LOG_FATAL, "%s", err)
			}
			exitCh <- err
		})
	} 
```

**Comment by ploxiln** [comment]
Not a bug :) just a bit tricky.

All the functions called within `exitFunc()` (for example `exitFunc(protocol.TCPServer(...))`) run until the program exits, then return a null error. Or, if they can't start up, then they exit with an error immediately. So if there is any non-nil error, it will be first. There might be more non-nil errors, but the just the first is logged, and only one error (nil or non-nil) is written to the unbuffered channel, because only one will be read.

In summary, there are two cases:
  * all components start successfully, all will later return `nil` on graceful program shutdown, nothing is logged and nothing is missed
  * one or more components fail to start and return a non-nil error immediately, and just the first one is logged

It would perhaps have been simpler to use a buffered channel with sufficient buffer size for all possible values, then all can be written without blocking even if just one is read ...
